components/PlanList/PlanList.js

```diff
@@ -1,28 +1,97 @@
-          <li>Medication Name :{plan.medication_name}</li>
-          <li>Active Ingredients :{plan.active_ingredients}</li>
-          <li>Dosage : {plan.dosage}</li>
-          <li>
-            Frequency : {plan.frequencyType} , {plan.frequencyTimes} , (
-            {plan.frequencyMornings ? "Mornings  " : " , "}
-            {plan.frequencyAfternoon ? "Afternoon  " : " , "}
-            {plan.frequencyEvening ? "Evenings  " : " , "}
-            {plan.frequencyBedTime ? "Bed Time  " : " "})
-          </li>
-          <li>Form : {plan.medicationForm}</li>
-          <li>Refill Date : {plan.refill}</li>
-          <li>Reason for taking : {plan.reason}</li>
-          <li>Instructions : {plan.instractions}</li>
-        </List>
-      ))}
+      <Title>My Medications Plan</Title>
+      <Span onClick={handleShowMyPlans}>
+        {showMyPlans ? "Hide Plans" : "Show Plans"} ({myPlans.length}) :
+      </Span>
+      {myPlans.length > 0 && showMyPlans && (
+        <>
+          {myPlans.map((plan, index) => (
+            <List key={index} importance={plan.importance}>
+              <li>Medication Name : {plan.medication_name}</li>
+              <li>Active Ingredients : {plan.active_ingredients}</li>
+              <li>Dosage : {plan.dosage}</li>
+              <li>
+                Frequency : {plan.frequencyType} , {plan.frequencyTimes} , (
+                {plan.frequencyMornings ? "Mornings  " : "   "}
+                {plan.frequencyAfternoon ? "Afternoon  " : "  "}
+                {plan.frequencyEvening ? "Evenings  " : "  "}
+                {plan.frequencyBedTime ? "Bed Time  " : "  "})
+              </li>
+              <li>Form : {plan.medicationForm}</li>
+              <li>Refill Date : {plan.refill}</li>
+              <li>Reason for taking : {plan.reason}</li>
+              <li>Instructions : {plan.instractions}</li>
+            </List>
+          ))}
+        </>
+      )}
+      {otherPlans.length > 0 && (
+        <>
+          <Title>Others Plans</Title>
+          {Object.values(
+            otherPlans.reduce((groupedPlans, plan) => {
+              const key = `${plan.name}-${plan.gender}-${plan.age}`;
+              if (!groupedPlans[key]) {
+                groupedPlans[key] = {
+                  name: plan.name,
+                  gender: plan.gender,
+                  age: plan.age,
+                  medications: [],
+                };
+              }
+              groupedPlans[key].medications.push(plan);
+              return groupedPlans;
+            }, {})
+          ).map((group, index) => (
+            <div key={index}>
+              <Heading>
+                {group.name} ({group.gender}, {group.age})
+              </Heading>
+              <Span onClick={() => handleShowPlan(index)}>
+                {showPlan === index ? "Hide Plans" : "Show Plans"} (
+                {group.medications.length}) :
+              </Span>
+              {showPlan === index &&
```

Comment:
since we now have carried over the key we could use that for the toggle function instead of the index. This is better because it does not rely on the order of elements.

This would mean that you ignore my advice above and instead call the state something like this:

```js
  const [activePlanKey, setActivePlanKey] = useState(null);
```

```js
const toggleActivePlan = (key) => {
    setActivePlanKey(activePlanKey === index ? null : index);
};
```

And then:
```suggestion
              <Span onClick={() => toggleActivePlan(key)}>
                {showPlan === index ? "Hide Plans" : "Show Plans"} (
                {group.medications.length}) :
              </Span>
              {activePlanKey === key &&
```

Reply:
i have tried to aply your Suggested changes but it shows me that key is not defined

Reply:
This is my complete file after implementing it:

```js
import { List, Title, Heading, Span } from "./PlanList.styled";
import { useState } from "react";

export default function PlanList({ plans }) {
  const myPlans = plans.filter((plan) => plan.for === "Me");
  const otherPlans = plans.filter((plan) => plan.for !== "Me");
  const [activePlanKey, setActivePlanKey] = useState(null);
  const [showMyPlans, setShowMyPlans] = useState(false); // Initialize state as false

  const handleActivePlanKey = (key) => {
    setActivePlanKey(activePlanKey === key ? null : key);
  };

  const handleShowMyPlans = () => {
    // Toggle showAllPlans state when heading is clicked
    setShowMyPlans(!showMyPlans);
  };
  console.log(plans);
  return (
    <>
      <Title>My Medications Plan</Title>
      <Span onClick={handleShowMyPlans}>
        {showMyPlans ? "Hide Plans" : "Show Plans"} ({myPlans.length}) :
      </Span>
      {myPlans.length > 0 && showMyPlans && (
        <>
          {myPlans.map((plan, index) => (
            <List key={index} importance={plan.importance}>
              <li>Medication Name : {plan.medication_name}</li>
              <li>Active Ingredients : {plan.active_ingredients}</li>
              <li>Dosage : {plan.dosage}</li>
              <li>
                Frequency : {plan.frequencyType} , {plan.frequencyTimes} , (
                {plan.frequencyMornings ? "Mornings  " : "   "}
                {plan.frequencyAfternoon ? "Afternoon  " : "  "}
                {plan.frequencyEvening ? "Evenings  " : "  "}
                {plan.frequencyBedTime ? "Bed Time  " : "  "})
              </li>
              <li>Form : {plan.medicationForm}</li>
              <li>Refill Date : {plan.refill}</li>
              <li>Reason for taking : {plan.reason}</li>
              <li>Instructions : {plan.instractions}</li>
            </List>
          ))}
        </>
      )}
      {otherPlans.length > 0 && (
        <>
          <Title>Others Plans</Title>
          {Object.values(
            otherPlans.reduce((groupedPlans, plan) => {
              const key = `${plan.name}-${plan.gender}-${plan.age}`;
              if (!groupedPlans[key]) {
                groupedPlans[key] = {
                  key,
                  name: plan.name,
                  gender: plan.gender,
                  age: plan.age,
                  medications: [],
                };
              }
              groupedPlans[key].medications.push(plan);
              return groupedPlans;
            }, {})
          ).map((group) => (
            <div key={group.key}>
              <Heading>
                {group.name} ({group.gender}, {group.age})
              </Heading>
              <Span onClick={() => handleActivePlanKey(group.key)}>
                {activePlanKey === group.key ? "Hide Plans" : "Show Plans"} (
                {group.medications.length}) :
              </Span>
              {activePlanKey === group.key &&
                group.medications.map((plan) => (
                  <List key={plan.id} importance={plan.importance}>
                    <li>Medication Name : {plan.medication_name}</li>
                    <li>Active Ingredients : {plan.active_ingredients}</li>
                    <li>Dosage : {plan.dosage}</li>
                    <li>
                      Frequency : {plan.frequencyType} , {plan.frequencyTimes} ,
                      ({plan.frequencyMornings ? "Mornings  " : "   "}
                      {plan.frequencyAfternoon ? "Afternoon  " : "   "}
                      {plan.frequencyEvening ? "Evenings  " : "   "}
                      {plan.frequencyBedTime ? "Bed Time  " : " "})
                    </li>
                    <li>Form : {plan.medicationForm}</li>
                    <li>Refill Date : {plan.refill}</li>
                    <li>Reason for taking : {plan.reason}</li>
                    <li>Instructions : {plan.instractions}</li>
                  </List>
                ))}
            </div>
          ))}
        </>
      )}
    </>
  );
}
```

Reply:
thank you for your helpful comment , i have already change it 👍